src/v3/src/util/theme.ts

```diff
@@ -10,116 +10,464 @@
  * See the License for the specific language governing permissions and limitations under the License.
  */
 
-import { ThemeOptions } from '@mui/material';
-import { Theme } from '@mui/material/styles/createTheme';
 import { odysseyTheme } from '@okta/odyssey-react-mui';
 import chroma from 'chroma-js';
-import { cloneDeep, merge } from 'lodash';
+import { set as _set, isUndefined } from 'lodash';
+import { buttonClasses } from "@mui/material/Button";
 
-import { BrandColors, LanguageDirection } from '../types';
+import { Brand } from '../types';
+import { DESIGN_TOKENS, DesignTokensType } from './designTokens';
+import { mergeThemes } from './mergeThemes';
 
-type DerivedTheme = {
-  primaryColor: string;
-  primaryColorLight: string;
-  primaryColorLightest: string;
-  primaryColorDark: string;
-  textColor: string;
-  inverseTextColor: string;
+type Palette = {
+  main: string;
+  light: string;
+  lighter: string;
+  dark: string;
+  contrastText: string;
 };
 
-const getInverseTextColor = (primaryColor: string): string => {
-  const contrastRatio = chroma.contrast(primaryColor, '#ffffff');
+type Theme = typeof odysseyTheme;
 
-  if (contrastRatio > 4.5) {
-    return '#ffffff';
-  }
-  return '#1d1d21';
-};
 
-export const deriveThemeFromBrand = (brand: BrandColors): DerivedTheme | null => {
-  try {
-    const isLightPrimaryColor = chroma(brand.primaryColor).get('hsl.l') > 0.24;
+const WHITE = '#ffffff';
+const BLACK = '#1d1d21';
+
+/**
+ * Determine whether to use BLACK (#1d1d21) or WHITE (#ffffff) based on a color.
+ * If WHITE vs color has a contrast ratio <4.5:1 this will return BLACK.
+ * Contrast ratios are symmetrical using the WCAG 2.x algorithm
+ *
+ * @param color color to compare against black/white
+ */
+const getInverseTextColor = (color: string): string => (
+  chroma.contrast(color, WHITE) > 4.5 ? WHITE : BLACK
+);
+
+/**
+ * Sets the design token to theme path iff the value is not undefined, i.e.,
+ * null, false, 0, etc. will be set. Only undefined will be ignored. Paths and
+ * values are not checked for type safety.
+ *
+ * @param obj {Theme} the theme object
+ * @param path {string} target path
+ * @param val the value to set, if defined.
+ */
+function set<T extends object, V>(obj: T, path: string | string[], val: V) {
+  return isUndefined(val) ? null : _set(obj, path, val);
```

Comment:
Do we need to use lodash's `isUndefined`?

Reply:
Replaced